uniffi_bindgen/src/templates/macros.rs

```diff
@@ -47,33 +52,63 @@
 {% macro to_rs_constructor_call(obj, cons) %}
 {% match cons.throws() %}
 {% when Some with (e) %}
-UNIFFI_HANDLE_MAP_{{ obj.name()|upper }}.insert_with_result(err, || -> Result<{{obj.name()}}, {{e}}> {
-    let _retval = {{ obj.name() }}::{% call to_rs_call(cons) %}?;
-    Ok(_retval)
-})
+    todo!("XXX - catch_unwind like below...");
+    let constructed = {{ obj.name() }}::{% call to_rs_call(cons) %}?;
+    let arc = std::sync::Arc::new(constructed);
+    Ok(std::sync:Arc::into_raw(arc))
 {% else %}
-UNIFFI_HANDLE_MAP_{{ obj.name()|upper }}.insert_with_output(err, || {
-    {{ obj.name() }}::{% call to_rs_call(cons) %}
-})
+    match std::panic::catch_unwind(|| {
+        {%- if obj.threadsafe() %}
+        let _new = {{ obj.name() }}::{% call to_rs_call(cons) %};
+        {%- else %}
+        let _new = std::sync::Mutex::new({{ obj.name() }}::{% call to_rs_call(cons) %});
+        {%- endif %}
+        let _arc = std::sync::Arc::new(_new);
+        std::sync::Arc::into_raw(_arc)
+    }) {
+        Ok(ptr) => {
+            *err = uniffi::deps::ffi_support::ExternError::default();
+            ptr as usize
+        },
+        Err(e) => {
+            *err = e.into();
+            0 as usize /*`std::ptr::null()` is a compile error */
+        }
+    }
 {% endmatch %}
 {% endmacro %}
 
+{% macro get_arc(obj) -%}
+    {% if obj.threadsafe() %}
+    let _arc = unsafe { std::sync::Arc::from_raw(ptr as *const {{ obj.name() }}) };
+    {% else %}
+    let _arc = unsafe { std::sync::Arc::from_raw(ptr as *const std::sync::Mutex<{{ obj.name() }}>) };
+    {% endif %}
+    // This arc now "owns" the reference but we need an outstanding reference still.
+    std::sync::Arc::into_raw(std::sync::Arc::clone(&_arc));
```

Comment:
I suggest using [`ManuallyDrop`](https://doc.rust-lang.org/std/mem/struct.ManuallyDrop.html) rather than leaking it back via `into_raw`, as I think it better captures the intent. If I had to write this code out by hand with lots of comments it would look something like:

```
// The pointed-to Arc is owned by the foreign-language code, so we mustn't drop it.
let _foreign_arc = ManuallyDrop::new(unsafe { Arc::from_raw(ptr as *const {{ obj.name() }}) });
// We want an owned copy of the Arc to pass internally to the Rust code.
let _arc = Arc::clone(&*_foreign_arc);

```